### PR TITLE
Add feature: Split VCF

### PR DESCRIPTION
**Pull request by spencerseale**
Adding a split feature for distributed split of a multi-sample VCF. A common obstacle occurring prior to VCF ingestion. 

https://app.shortcut.com/tiledb-inc/story/48501/get-the-multisample-vcf-splitter-released-as-a-public-registered-task-graph

### CODE REVIEW

File: src/tiledb/cloud/vcf/split.py [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+
+    graph = split_vcf(
+        vcf_uri="s3://tiledb-spencer/projects/horses/raw/thesis_intersect_pub.decomposed.vcf.gz",
+        output_uri="s3://tiledb-spencer/projects/vcf-isolate-dev/isolated-singles",
+        config={"vfs.s3.region": "us-west-2"},
+        samples=["A1543"],
+        namespace="TileDB-Inc",
+        acn="tiledb-cloud-sandbox-role",
+        compute=True,
+    )
```
Comment (gspowley) [inline src/tiledb/cloud/vcf/split.py:343]: Please remove the testing code. It could be added to the manually run pytests, see https://github.com/TileDB-Inc/TileDB-Cloud-Py/blob/main/tests/test_vcf.py.

File: src/tiledb/cloud/vcf/split.py [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+
+    Args:
+        vcf_uri:
+            S3 path to aggregate VCF.
+        config:
+            TileDB config params.
+
+    Returns:
+        Samples included in VCF.
+    """
```
Comment (gspowley) [inline src/tiledb/cloud/vcf/split.py:22]: Does this comment style work with the doc generator? I know we use sphinx style in other places.

File: src/tiledb/cloud/vcf/split.py [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+    vcf_uri: str,
+    output_uri: str,
+    namespace: str,
+    acn: str,
+    resources: Mapping[str, str] = {"cpu": "2", "memory": "30Gi"},
+    compute: bool = True,
+    verbose: bool = False,
+    samples: Optional[Sequence[str]] = None,
+    retry_count: int = 1,
+    max_workers: int = 100,
```
Comment (gspowley) [inline src/tiledb/cloud/vcf/split.py:207]: These params need docstrings.

**Review by gspowley — APPROVED** [review]
Looks good, a few nits. Adding a manually run pytest is a good idea.

File: src/tiledb/cloud/vcf/split.py (hunk 1 above)
Comment (spencerseale) [reply src/tiledb/cloud/vcf/split.py:343]: Yeah I will! This PR is still draft before ready for approval. i'll add unit tests as well

File: src/tiledb/cloud/vcf/split.py (hunk 2 above)
Comment (spencerseale) [reply src/tiledb/cloud/vcf/split.py:22]: Sphinx picks it up nicely, but you're right I should assimilate with other doc style and was just a force of habit

**Comment by spencerseale** [comment]
@gspowley @leipzig reformatted doc str to align with rest of base and confirmed local doc build looks good. Created `tests/vcf` and added unit tests for this module and moved other tests to same module. Ready to for review

File: src/tiledb/cloud/vcf/split.py (hunk 2 above)
Comment (spencerseale) [reply src/tiledb/cloud/vcf/split.py]: Quarto did not like that formatting, thank you for noticing this, should be fixed now

**Review by JohnMoutafis — APPROVED** [review]
lgtm.
Please add a description on the PR and merge it.

**Review by leipzig — CHANGES_REQUESTED** [review]
I think this is great. Let's just make sure you have tests for the edge cases
- splitting a VCF with 1 sample
- splitting a VCF with 0 samples
- splitting a VCF with 100 samples but no variants
These are all "do nothing, but notify" edge cases

**Comment by spencerseale** [comment]
@leipzig been thinking about the edge cases:

splitting a VCF with 1 sample: 
- If the user passes in the single sample, then bcftools will perform split without error and logging that. Doing exactly what was requested. I don't believe we should try to engineer something to ensure they're not confused, but rather ensure we do explicitly whats requested to caller.
- If user doesn't specify samples, the first stage of task graph will log that just one sample exists in VCF and the second stage will split that one sample. It will be clear from first stage the number of samples in VCF.

splitting a VCF with 0 samples:
- If they don't specify samples, will perform first stage where samples are listed and dynamically spawn subsequent stage based on number of values in samples. So no splitting will occur and the first stage would list that no samples were found.
- If they do specify a sample that doesn't exist in the vcf aggregate, I made an edit to promote the bcftools error and that will be returned to logger and look like: `Error: subset called for sample that does not exist in header`.

splitting a VCF with 100 samples but no variants:
- Would just split each sample as it's own node and return headers without variant data. Since we would still have to start a node per sample to get this check, I don't see any ramp in efficiency/cost savings if we first checked for variants. To check for variant data, I would need to then open each of the split VCF files once written to s3, incurring additional billing (although small).

Let me know your comments on these. My opinion is because the logger is verbose enough and i've promoted bcftools to return stderr when it returns a non-zero return code that the tests are sufficient.

**Comment by spencerseale** [comment]
I know you're busy so no rush on this PR, just when you'd like it to be pushed. But do I need to do anything academy-wise for this prior to its release?

**Comment by leipzig** [comment]
> I know you're busy so no rush on this PR, just when you'd like it to be pushed. But do I need to do anything academy-wise for this prior to its release?

should be documented at the end of  life sciences > popgen > tutorials > advanced > Scalable Ingestion

**Review by leipzig — APPROVED** [review]
